pkg/api/controllers/catalog/catalog.go

```diff
@@ -0,0 +1,54 @@
+package catalog
+
+import (
+	"context"
+
+	helmlib "github.com/rancher/rancher/pkg/catalog/helm"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+func Register(ctx context.Context, context *config.ScaledContext) {
+	controller := context.Management.Catalogs("").Controller()
+	controller.AddHandler(ctx, "catalog", destroyCatalogSync)
+
+	clusterCatalogController := context.Management.ClusterCatalogs("").Controller()
+	clusterCatalogController.AddHandler(ctx, "clusterCatalog", destroyClusterCatalogSync)
+
+	projectCatalogController := context.Management.ProjectCatalogs("").Controller()
+	projectCatalogController.AddHandler(ctx, "projectCatalog", destroyProjectCatalogSync)
+}
+
+func destroyCatalogSync(key string, obj *v3.Catalog) (runtime.Object, error) {
+	if obj == nil {
+		return destroy(key)
```

Comment:
There's no reason guarantee that you will see the delete of an object. What is the impact if destroy is never called?

Reply:
If destroy is never called, and another catalog is created with the same namespace & name but different URL, then the server may analyze and use the wrong catalog cache.

Reply:
We could verify the URL on catalog cache initialization and recreate the cache if needed. Then the only side effect of not calling destroy would be leaving some cache on disk if the catalog is deleted.

Reply:
This PR has been updated to scan through the cache directory and remove any directories which do not correspond with a catalog, it will clean up stale items in the cache once at start, on a period of every hour, and catalog changes are debounced to clean up after a minute of inactivity.